### PR TITLE
feat: cache and store content claims

### PR DESCRIPTION
**Pull request by alanshaw**
This PR changes the `ClaimLookup` interface to allow caching and permanent storage (publishing) of content claims. It also allows retrieval by content claim root CID.

### CODE REVIEW

File: pkg/blobindex/shardeddagindex.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func View(root ipld.Link, blockMap map[ipld.Link]ipld.Block) (ShardedDagIndexVie
 	if shardedDagIndexData.DagO_1 == nil {
 		return nil, NewUnknownFormatError(fmt.Errorf("unknown index version"))
 	}
-	dagIndex := NewShardedDagIndexView(root, len(shardedDagIndexData.DagO_1.Shards))
+	dagIndex := NewShardedDagIndexView(shardedDagIndexData.DagO_1.Content, len(shardedDagIndexData.DagO_1.Shards))
```
Comment (alanshaw) [inline pkg/blobindex/shardeddagindex.go:80]: Bug where a decoded blob index content field was being set to the CID of the CAR.

File: pkg/construct/construct.go [hunk 2]
```diff
@@ -255,6 +255,7 @@ func Construct(sc ServiceConfig, opts ...Option) (Service, error) {
 
 		s.startupFuncs = append(s.startupFuncs, func(context.Context) error { jq.Startup(); return nil })
 		s.shutdownFuncs = append(s.shutdownFuncs, jq.Shutdown)
+		cachingQueue = jq
```
Comment (alanshaw) [inline pkg/construct/construct.go:258]: Bug `cachingQueue` remained `nil`.

File: pkg/service/service.go [hunk 3]
```diff
@@ -282,7 +283,7 @@ func (is *IndexingService) Get(ctx context.Context, claim ipld.Link) (delegation
 // ideally however, IPNI would enable UCAN chains for publishing so that we could publish it directly from the storage service
 // it doesn't for now, so we let SPs publish themselves them direct cache with us
 func (is *IndexingService) Cache(ctx context.Context, provider peer.AddrInfo, claim delegation.Delegation) error {
-	return Cache(ctx, is.blobIndexLookup, is.claims, is.providerIndex, is.provider, claim)
+	return Cache(ctx, is.blobIndexLookup, is.claims, is.providerIndex, provider, claim)
```
Comment (alanshaw) [inline pkg/service/service.go:286]: Bug - we were using the indexer provider info when caching claims from storage nodes 🤦

File: pkg/metadata/metadata.go [hunk 4]
```diff
@@ -99,6 +92,10 @@ func init() {
 	mdctx = mdctx.WithProtocol(EqualsClaimID, func() ipnimd.Protocol { return &EqualsClaimMetadata{} })
 	mdctx = mdctx.WithProtocol(LocationCommitmentID, func() ipnimd.Protocol { return &LocationCommitmentMetadata{} })
 	MetadataContext = mdctx
+	// WithProtocol creates a _new_ context. Assign it to Default so claim
+	// metadata can be decoded elsewhere by refrencing libipni metadata.Default
+	// i.e. ipni-publisher
+	ipnimd.Default = mdctx
```
Comment (alanshaw) [inline pkg/metadata/metadata.go:98]: Bug - see comment. The `ipni-publisher` store uses `metadata.Default` and was not able to decode our custom metadatas. I think ideally we'd pass metadata context through?

File: pkg/metadata/metadata.go [hunk 5]
```diff
@@ -85,12 +84,6 @@ const EqualsClaimID = 0x3E0001
 // LocationCommitmentID is the multicodec for location commitments
 const LocationCommitmentID = 0x3E0002
 
-var nodePrototypes = map[multicodec.Code]schema.TypedPrototype{
-	IndexClaimID:         indexClaimMetadata,
-	EqualsClaimID:        equalsClaimMetadata,
-	LocationCommitmentID: locationCommitmentMetadata,
-}
```
Comment (alanshaw) [inline pkg/metadata/metadata.go:92]: Bug - These were `nil` as `init` has not run yet.

File: pkg/construct/construct.go (hunk 2 above)
Comment (hannahhoward) [reply pkg/construct/construct.go:258]: #oops

File: deploy/s3.tf [hunk 6]
```diff
@@ -47,4 +47,8 @@ resource "aws_s3_bucket_policy" "ipni_store_policy" {
 
 resource "aws_s3_bucket" "notifier_head_bucket" {
   bucket = "${terraform.workspace}-${var.app}-notifier-head-bucket"
-}
\ No newline at end of file
+}
+
+resource "aws_s3_bucket" "claim_store_bucket" {
```
Comment (hannahhoward) [inline deploy/s3.tf:52]: Question: in prod, shouldn't this be the same bucket we already have to legacy claims? Or do we store it different.

File: pkg/metadata/metadata.go (hunk 4 above)
Comment (hannahhoward) [reply pkg/metadata/metadata.go:98]: yea we should

File: pkg/service/service.go [hunk 7]
```diff
@@ -28,62 +29,28 @@ import (
...
-	// other multihashes in the index
-	Find(ctx context.Context, contextID types.EncodedContextID, provider model.ProviderResult, fetchURL url.URL, rng *metadata.Range) (blobindex.ShardedDagIndexView, error)
-}
-
 // IndexingService implements read/write logic for indexing data with IPNI, content claims, sharded dag indexes, and a cache layer
 type IndexingService struct {
-	blobIndexLookup BlobIndexLookup
-	claimLookup     ClaimLookup
-	providerIndex   ProviderIndex
+	blobIndexLookup blobindexlookup.BlobIndexLookup
```
Comment (hannahhoward) [inline pkg/service/service.go:44]: Apologies for my GO related thinking here -- there's a meme in go of declare the interface where you use it. Honestly, not sure I've ever fully bought in :)

File: pkg/service/service.go (hunk 3 above)
Comment (hannahhoward) [reply pkg/service/service.go:286]: Fle under: oops.

**Review by hannahhoward — CHANGES_REQUESTED** [review]
Generally awesome and LGTM

One big caveat, which I realize is actually my own mistake. I extracted the use of metadata to the go-metadata package but failed to delete the actual metadata file. OOPS -- we should do that right away to avoid confusion

**Comment by alanshaw** [comment]
> One big caveat, which I realize is actually my own mistake. I extracted the use of metadata to the go-metadata package but failed to delete the actual metadata file. OOPS -- we should do that right away to avoid confusion

Ah, damn, yes ok that explains a lot. I thought I'd seen those issues before 🤦

**Review by hannahhoward — APPROVED** [review]

File: pkg/service/service.go (hunk 7 above)
Comment (alanshaw) [reply pkg/service/service.go:44]: i_have_no_idea_what_im_doing

**Pull request merged**